src/librustc/session/mod.rs

```diff
@@ -864,6 +864,11 @@ impl Session {
     pub fn teach(&self, code: &DiagnosticId) -> bool {
         self.opts.debugging_opts.teach && !self.parse_sess.span_diagnostic.code_emitted(code)
     }
+
+    /// Are we allowed to use features from the Rust 2018 epoch?
+    pub fn rust_2018(&self) -> bool {
```

Comment:
I think we should prepare for macro compatibility here, and make this some kind of test that takes a span. I'm not sure where the method should go -- maybe put it on the tcx for now? That seems like it would *definitely* have access to whatever data it needs to make the determination.

Reply:
I don't think the tcx makes sense; the idea behind epoches was that most of the changes would be to stuff like parsing and all, the "language frontend". We don't have a tcx when parsing. I think we should keep this unresolved for now and when someone knowledgeable about macro stuff comes along they can look into this.

I _think_ the Session already has a lot of the Span info that can be used here; at least "originates from macro from X crate" can be done from the Session, though the interaction of many macros may be confusing.

Reply:
> I don't think the tcx makes sense

In the future, we plan to move so that the tcx is the first thing that is created, so that everything can be query driven.  This is why I suggested tcx. However, session is also fine. 

> I think we should keep this unresolved for now and when someone knowledgeable about macro stuff comes along they can look into this.

Mm, I'm not sure. I think it'll be useful to require a span so that, for any change, we are at least thinking about how to deal with it. 

--

Eh. Having thought about it some more, while I'd rather lay out a "forwards compatible" API with what we will likely eventually need, I don't care enough to block on it. And there's always an argument for doing the simplest possible thing to start.